### PR TITLE
fix sdpa incorrect early return on 0 head dim qk w/ valid v

### LINKED ISSUE #184330: SDPA returns 0 output and 0 gradient with corner case input since 2.12
### 🐛 Describe the bug

We meet a corner case bug for SDPA produced by the tiny Qwen2.5-VL vision model: In our config, attention after RoPE: hidden_size=16, num_heads=16 -> head_dim=1, while RoPE cos/sin have last dimension 0. As a result q/k become (..., 0), but v remains (..., 1). With this input SDPA will return 0 output and 0 gradient, while with eager mode it is OK. w/ PT 2.11, both sdpa and eager attn_implementation are OK. So is it a bug introduced by torch 2.12, or it is an expected behavior?
```py
import torch
import torch.nn.functional as F


def eager_attention(query, key, value, scale):
    scores = torch.matmul(query, key.transpose(-2, -1)) * scale
    probs = torch.softmax(scores, dim=-1, dtype=torch.float32).to(query.dtype)
    return torch.matmul(probs, value)


def run(use_sdpa, device):
    torch.manual_seed(0)
    query = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
    key = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
    value = torch.randn((1, 16, 16, 1), device=device, requires_grad=True)
    grad_output = torch.randn((1, 16, 16, 1), device=device)

    if use_sdpa:
        output = F.scaled_dot_product_attention(
            query,
            key,
            value,
            dropout_p=0.0,
            is_causal=False,
            scale=1.0,
        )
    else:
        output = eager_attention(query, key, value, scale=1.0)

    output.backward(grad_output)
    return output.detach().abs().sum().item(), value.grad.detach().abs().sum().item()


def compare(device):
    sdpa_output_sum, sdpa_value_grad_sum = run(use_sdpa=True, device=device)
    eager_output_sum, eager_value_grad_sum = run(use_sdpa=False, device=device)

    print(f"\n{device}")
    print(f"sdpa : output_abs_sum={sdpa_output_sum:.9e}, value_grad_abs_sum={sdpa_value_grad_sum:.9e}")
    print(f"eager: output_abs_sum={eager_output_sum:.9e}, value_grad_abs_sum={eager_value_grad_sum:.9e}")

    if sdpa_value_grad_sum == 0.0 and eager_value_grad_sum > 0.0:
        print("BUG: SDPA returns zero value gradient while eager attention returns a non-zero gradient.")


def main():
    print(f"torch: {torch.__version__}")
    compare("cuda")
    compare("cpu")


if __name__ == "__main__":
    main()
```

### Versions

```
PyTorch version: 2.12.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar  3 2026, 12:15:18) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.292-1.el8.elrepo.x86_64-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe

Nvidia driver version: 570.133.20
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.20.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
BIOS Vendor ID:                     Intel(R) Corporation
Model name:                         Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
BIOS Model name:                    Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz  CPU @ 2.6GHz
BIOS CPU family:                    179
CPU family:                         6
Model:                              106
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           6
Frequency boost:                    enabled
CPU(s) scaling MHz:                 97%
CPU max MHz:                        3400.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           80 MiB (64 instances)
L3 cache:                           96 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Vulnerable: No microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] apollo-torch==1.0.3
[pip3] galore-torch==1.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.20.1
[pip3] onnx-ir==0.2.0
[pip3] onnxscript==0.6.2
[pip3] torch==2.12.0+cu126
[pip3] torchdata==0.11.0
[pip3] torchvision==0.27.0+cu126
[pip3] transformer_engine_torch==2.13.0
[pip3] triton==3.7.0
```

cc @drisspg @liangel-02 @howardzhang-cv

### PR DESCRIPTION
**Pull request by liangel-02**
fixes #184330

today in attention.cpp, we have this check in `scaled_dot_product_attention`: 
https://github.com/pytorch/pytorch/blob/93c10b22b50317447f2ba046680a57c4c8681d4a/aten/src/ATen/native/transformers/attention.cpp#L739-L746

which guards against empty outputs by checking if any of q/k/v has numel()==0. however, this is too strict since even if head dim of q/k = 0 but v still has valid elements, then attention output is still defined (should just be the mean of all values). 

the guard should actually just check if any of v's dims are 0 (which checks for empty batch, head dims, 0 kv seqlen, and 0 v headdim) and if q seqlen = 0. 

note that in the head_dim=0 case, you will end up using math backend since flash/cudnn doesnt support this

### CODE REVIEW

**Review by drisspg — COMMENTED** [review]
can you update the sdp_utils so that if we ar eint eh scenario where query and K have dim 0 they go to the math backend I dont think the other ones
 handle it

File: test/test_transformers.py [hunk 1]
```diff
@@ -2212,25 +2212,59 @@ def test_scaled_dot_product_attention_fp16_overflow(self, device):
...
+
+    @parametrize("value_dim", [1, 8])
+    def test_sdpa_zero_qk_head_dim_matches_eager_attention(self, device, value_dim):
+        torch.manual_seed(0)
+        query = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
+        key = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
+        value = torch.randn((1, 16, 16, value_dim), device=device, requires_grad=True)
+        grad_output = torch.randn((1, 16, 16, value_dim), device=device)
+
+        actual = F.scaled_dot_product_attention(query, key, value, dropout_p=0.0, scale=1.0)
```
Comment (drisspg) [inline test/test_transformers.py]: Try with scale = None

File: test/test_transformers.py [hunk 2]
```diff
@@ -2212,25 +2212,59 @@ def test_scaled_dot_product_attention_fp16_overflow(self, device):
...
+        query_ref = query.detach().clone().requires_grad_()
+        key_ref = key.detach().clone().requires_grad_()
+        value_ref = value.detach().clone().requires_grad_()
+        scores = torch.matmul(query_ref, key_ref.transpose(-2, -1))
+        probs = torch.softmax(scores, dim=-1, dtype=torch.float32).to(query_ref.dtype)
+        expected = torch.matmul(probs, value_ref)
+        expected.backward(grad_output)
+
+        self.assertEqual(actual, expected)
+        self.assertEqual(actual_value_grad, value_ref.grad)
```
Comment (drisspg) [inline test/test_transformers.py:2272]: can you triple check query and key grads

**Review by drisspg — APPROVED** [review]
